### PR TITLE
Implement IPC shutdown

### PR DESCRIPTION
**Pull request by parsonsmatt**
## Description

This PR implements IPC shutdown for the Node API.

## Type of change

- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🛠 New feature (non-breaking change which adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [ ] 🔨 New or improved tests for existing code
- [ ] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist

- [ ] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [ ] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [ ] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.
- [ ] CHANGELOG entry has been added and is linked to the correct PR on GitHub.

## Testing checklist

- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

## How to merge

Send the message `bors r+` to merge this PR. For more information, see
[`docs/how-to/bors.md`](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/how-to/bors.md).

### CODE REVIEW

File: node/src/Cardano/Node/API.hs [hunk 1]
```diff
@@ -232,102 +239,20 @@ instance Core.HasSlottingVar SettingsCtx where
...
 --
--- NOTE (legacy): 'postponeUpdate', "Pos.Wallet.Web.Methods.Misc".
-postponeUpdate :: MonadIO m => Kernel.PassiveWallet -> m ()
-postponeUpdate w = update' (w ^. Kernel.wallets) $ RemoveNextUpdate
-
--}
+-- TODO: verify this is a real thought and not, in fact, bad
+postponeUpdate :: Handler NoContent
+postponeUpdate = do
+    pure NoContent
```
[the remark this answers is not in the document]
Comment (KtorZ) [reply node/src/Cardano/Node/API.hs]: Yeah, was about to mention it on the `applyUpdate`; we should still be tracking updates in plain memory. Some sort of `MVar` or `IORef` in the node context. So the logic is similar as what we had before, we just don't persist the state.

[the remark this answers is not in the document]
Comment (parsonsmatt) [reply node/src/Cardano/Node/API.hs]: What is the logic, though? From my perspective, the node knows about a new update. When we ask "What's the next update?", then the node responds with the update if available. To update, you just restart the node software. To postpone, you just... don't restart te node software, which is business-as-usual.

[the remark this answers is not in the document]
Comment (KtorZ) [reply node/src/Cardano/Node/API.hs]: The current logic is as follows:

- We call `s` the queue of available updates (FIFO), maintained in memory. 
- When a new update is available, the node stores it in an `MVar ConfirmedProposalState` `ucDownloadedUpdate`
- Until now, the wallet was polling this MVar regularly to catch the update and push ii on top of the queue `s`
- When  `GET /api/v1/next-update`, the node will:
    - Clean up `s` by removing all updates that aren't relevant anymore (cf `isActualUpdate` in `Cardano.Wallet.Kernel.DB.Updates`).
    - Peek at the last update on top of `s`
    - If `s` isn't empty, returns the corresponding software version mentioned in the update alongside a 200 response.
    - If `s` is empty, the node returns 404. 

- When `POST /api/v1/postpone-update`, the node remove the oldest update from the the queue `s`. No matter the state of the queue, the node returns a `204 No Content`.

- When `POST  /api/v1/apply-update`, the node remove the oldest update from the queue and no matter the state of the queue, restarts the node with an exit code `20`.

A few interesting things to know (cf `Pos.Network.Update.Download`)

- The last downloaded update is applied, no matter what, when the node restart. The exit code `20` is just a mechanism to tell the underlying launcher that the node should be restarted. But applying the update happens anyway if the node is restarted (after a sigkill, or a clean shutdown). 

- The node doesn't actually download more than one update at a time. 
    - If `ucDownloadedUpdate :: MVar ConfirmedProposalState` hasn't been emptied, it displays a log "We won't download an update for proposal [...]" and ignores the proposal.
    - If `ucDownloadedUpdate` is empty, the node checks whether there's already a file on disk with the update. It throws if there's one. 

---

So basically, there can be only one pending update, and that's the one stored in `ucDownloadedUpdate`. I am afraid that the existing code is sort-of-broken. By emptying this `MVar` early, we are exposing the node to potential crashes... Beside, this makes the underlying queue completely superfluous since there can only be one element in that queue! Perhaps a better way would be to just have another `MVar` telling us whether we've already accounted for the other `ucDownloadedUpdate`, without emptying it. Those `MVar` will be discarded anyway when restarting.

**Comment by rvl** [comment]
Bors is waiting for macOS builds. We must fix that.

bors r+

**Comment by rvl** [comment]
Updated because cross-compilation PR was merged.

bors r+
